FitGoal/FitGoal/Singup&Singin/SocialMediaAuthenticator.swift

```diff
@@ -169,25 +141,41 @@ class SocialMediaAuthenticator: NSObject, GIDSignInDelegate {
         }
     }
     
-    // MARK: -Google Delegate
-    func sign(_ signIn: GIDSignIn!, didSignInFor user: GIDGoogleUser!, withError error: Error!) {
-        if let error = error {
-            userInfoCallback?(.failure(error))
+    private func authenticateUser(with authenticationMethod: AuthenticationMethod, completion: @escaping UserInfoCallback) {
+        let credentials = authenticationMethod.credentials
+        Auth.auth().signIn(with: credentials) { (dataResults, error) in
+            if let error = error {
+                self.handleLoginError(error: error, authenticationMethod: authenticationMethod, completion: completion)
+            } else {
+                self.parseUserInformation(from: dataResults, error: error, completion: completion)
+            }
+        }
+    }
+    
+    private func handleLoginError(error: Error, authenticationMethod: AuthenticationMethod, completion: @escaping UserInfoCallback) {
+        let nsError = error as NSError
+        if nsError.code == AuthErrorCode.accountExistsWithDifferentCredential.rawValue {
+            self.handleAccountExistsWithDifferentCredentials(error: nsError, authenticationMethod: authenticationMethod, completion: completion)
         } else {
-            guard let authentication = user.authentication else { return }
-            
-            guard let email = user.profile.email else {
-                userInfoCallback?(.failure(MissingUserInfoError.noEmailFound))
-                return
+            completion(.failure(error))
+        }
+    }
+    
+    private func handleAccountExistsWithDifferentCredentials(error: NSError, authenticationMethod: AuthenticationMethod, completion: @escaping UserInfoCallback) {
+        let email = error.userInfo[AuthErrorUserInfoEmailKey] as! String
```

Comment:
Nothing ensures that key has to always exist in the error, so please don't use the force unwrap.